lib/portlayer/event/manager.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package event
+
+import (
+	"reflect"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/vmware/vic/lib/portlayer/event/collector"
+	"github.com/vmware/vic/lib/portlayer/event/events"
+)
+
+type Manager struct {
+	cos  collectorCache
+	subs subscriberCache
+}
+
+type collectorCache struct {
+	mu sync.RWMutex
+
+	collectors map[string]collector.Collector
+}
+
+type subscriberCache struct {
+	mu sync.RWMutex
+
+	subscribers map[interface{}]map[string]func(events.Event)
+}
+
+func NewEventManager(collectors ...collector.Collector) *Manager {
+	mgr := &Manager{
+		cos: collectorCache{
+			collectors: make(map[string]collector.Collector),
+		},
+		subs: subscriberCache{
+			subscribers: make(map[interface{}]map[string]func(events.Event)),
+		},
+	}
+
+	// register any collectors provided
+	for i := range collectors {
+		mgr.RegisterCollector(collectors[i])
+	}
+
+	return mgr
+}
+
+func (mgr *Manager) RegisterCollector(collector collector.Collector) {
+	if collector == nil {
+		return
+	}
+
+	mgr.cos.mu.Lock()
+	defer mgr.cos.mu.Unlock()
+
+	collector.Register(mgr.Publish)
+
+	mgr.cos.collectors[collector.Name()] = collector
```

Comment:
Currently I don't think it's an issue as we are only getting VM events -- but I agree that it might need to be changed to allow multiple collectors of type vsphere event collector.  I'd prefer to tackle it in a subsequent change if that's agreeable.